### ISSUE TITLE
Variable hydrolysis unit and max

### ISSUE THREAD

**Issue by curzon01** [alexdelprete/HA-NeoPool-MQTT#2]
Find a way to get hydrolysis unit from SENSOR $.Neopool.Hydrolysis.Unit. The current fix !g/h" doesn't work for % systems (like my).

Is there a way to get value for `unit_of_measurement` from MQTT?

Also, the max value does not match (currently set at 33 g/h, as % systems require 100, in general lwe need SENSOR $.Neopool.Hydrolysis.Max)

**Comment by alexdelprete** [comment]
Unfortunately `unit_of_measurement` and `max` do not support templates. :(

What I can do is make it generic, so a g/H (%) unit and 0-100 min max. When I find a proper solution for HA, we'll fix it. What do you think?

**Comment by curzon01** [comment]
> What I can do is make it generic, so a g/H (%) unit and 0-100 min max. When I find a proper solution for HA, we'll fix it.

as NPHydrolysis reverts higher value than system allowed with an error, go this way

**Comment by alexdelprete** [comment]
I also opened a request on HA forums, hope some dev there will answer: https://community.home-assistant.io/t/mqtt-number-unit-of-measurement-and-max-values-from-topic-json/633249

Even though, from what I read, it's not supported. So the only thing would be to ask for a FR.

**Comment by alexdelprete** [comment]
> as NPHydrolysis reverts higher value than system allowed with an error, go this way

ok. what do I use for unit_of_measurement? proposals:

```
unit_of_measurement: % (g/h)
unit_of_measurement: % - g/h
unit_of_measurement: % / g/h

```

**Comment by curzon01** [comment]
BTW I can, instead reverting invalid high values, cut it to Mac allowed an set this. So if anyone wants to set 100 on your system, it will set the max if your system.

**Comment by alexdelprete** [comment]
> BTW I can, instead reverting invalid high values, cut it to Mac allowed an set this. So if anyone wants to set 100 on your system, it will set the max if your system.

That would be GREAT.

**Comment by curzon01** [comment]
% / g/h

**Comment by alexdelprete** [comment]
> BTW I can, instead reverting invalid high values, cut it to Max allowed an set this. So if anyone wants to set 100 on your system, it will set the max of your system.

I thought about it today...and came up with a proper solution that would work for all devices:

Assumptions/Facts:

- we have devices that work in % mode and others in g/h
- from what I read, Hayward devices are all in g/h, Sugar Valley in % mode
- % mode is the smartest method of the two
- neopool driver knows in what mode the device is working
- neopool driver knows the cell max g/h production
- HA is not flexible enough to allow unit_of_measurement and max to be dynamically configured

My proposed solution is that NeoPool driver manages % mode for ALL devices and HA is configured in % mode: for devices configured in g/h mode, NeoPool driver would translate the % set in g/h.

This means that NPHydrolysis command only accepts % value, and the driver translates that based on the device working mode.

Basically, we would abstract the diversity of the device configuration. Let me know what you think.

**Comment by curzon01** [comment]
What is the content of your MBF_PAR_HIDRO_NOM = 0x0306 register, is this 330 (`Backlog NPResult 0;NPRead 0x306`)?

**Comment by alexdelprete** [comment]
> What is the content of your MBF_PAR_HIDRO_NOM = 0x0306 register

Yes, my cell max production is 33g/h:

```console
09:13:09.527 CMD: Backlog NPResult 0;NPRead 0x306
09:13:09.555 MQT: stat/SmartPool/NPRESULT = {"NPResult":0}
09:13:09.802 MQT: stat/SmartPool/NPREAD = {"NPRead":{"Address":774,"Data":330}}
```

> If so, is your cell capable of delivering 33 g/h, or is this the device setting and your cell has a lower output?

No, I bought that specific model (33g/h) because I have an 85m3 pool, and I sized the cell to the double of the minimum requirement.

What are your thoughts about my proposal? Or are you already implementing it? :)

**Comment by curzon01** [comment]
I wanted to have the info from a g/h system before.

The idea is good, yes I think I will implement that.

For info, I can output these values (max g/h and current g/h) additionally for g/h systems, which you can then ignore.

**Comment by curzon01** [comment]
BTW: I have updated the [NeoPool documentation](https://tasmota.github.io/docs/NeoPool/) yesterday (added the JSON key desc and M5Stack usage).

**Comment by alexdelprete** [comment]
> The idea is good, yes I think I will implement that (then also the description is correct again :-) ).

That is excellent news Norbert. So happy to hear that. I'm working on this stuff for quite some days and I really want it to be as good as it can be. :)

The description will be something like: `xx% (yy g/h)`.

> For info, I can output these values (max g/h and current g/h) additionally for g/h systems, which you can then ignore

You already implemented the max g/h (Hydrolysis.Max) and production g/h (Hydrolysis.Data) is already there. Am I missing something? :)

```json
        "Hydrolysis": {
            "Data": 0,       // NUMBER - Cell Production Value (0-100)
            "Unit": "g/h",   // (unit % or g/h depending on device/manufacturer)
            "Setpoint": 18,  // SENSOR - Production Measurement
            "Max": 33,       // SENSOR - Max Cell Production Value
```

**Comment by alexdelprete** [comment]
> This is totally mixed and % or g/h setting is a pure marketing decision of the respective distributors, which can not be determined by the brand name alone.

Good to know. But since we're smarter than marketing guys, we came up with a good solution to address this chaos properly. Everyone will use % mode, thanks to NeoPool driver, that's a great value proposition. :)

BTW: did you see the brand new UI? ;)

**Comment by curzon01** [comment]
> You already implemented the max g/h (Hydrolysis.Max) and production g/h (Hydrolysis.Data) is already there. Am I missing something? :)

Yeah, "Data" currently delivers g/h on your system and will deliver 0-100% in future also on your system.

I want do keep both and enhance the info like that

```JSON
        "Hydrolysis": {
            "Data": 100,         // always 0-100
            "Unit": "%",         // always "%"
            "Setpoint": 50,      // always in %
            "Max": 100,          // always 100
            "Systemunit": "g/h", // Only avail. on g/h systems
            "Systemmax": 33,     // Only avail. on g/h systems
```

the keys "Unit" and "Max" are then actually superfluous, but would like to keep this for backward compatibility

**Comment by curzon01** [comment]
> BTW: did you see the brand new UI? ;)

Yep, just briefly, I also have customers ;-)

**Comment by curzon01** [comment]
>  did you see the brand new UI? ;)

did the same as before with your ha-neopool-mqtt-lovelace.yaml (dashboard config raw edit and paste) bnut got an error:

 r: At path: views -- Expected an array value, but received: undefined

**Comment by alexdelprete** [comment]
> did the same as before with your ha-neopool-mqtt-lovelace.yaml (dashboard config raw edit and paste) bnut got an error:

the new UI relies on some external lovelace cards to be installed through HACS. Did you read that in the updated readme? Do it when you have taken care of customers, takes 10m. :)

**Comment by curzon01** [comment]
> Did you read that in the updated readme? 

I'd simulate a user, no :see_no_evil: :grinning: :rofl:

**Comment by alexdelprete** [comment]
> the keys "Unit" and "Max" are then actually superfluous, but would like to keep this for backward compatibility

But I didn't see the measured g/h data in that json snippet. How about this?
`Unit` and `Max` are really superfluous. :)

```json
  "Hydrolysis": {
    "Data": {
        "Percent": "50",   // measured value (0-100)
        "GramsHour": "16"  // calculated value based on Data.Percent of Systemmax
    },
    "Unit": "%",         // always "%"
    "Setpoint": 50,      // always in %
    "Max": 100,          // always 100
    "SystemUnit": "g/h", // Only avail. on g/h systems
    "SystemMax": 33,     // Only avail. on g/h systems
```

**Comment by alexdelprete** [comment]
> > Did you read that in the updated readme?
> 
> I'd simulate a user, no 🙈 😀 🤣

RTFM doesn't apply to experienced devs. But it should...🤣

**Comment by curzon01** [comment]
No, this is a breaking change, suggestion:

```JSON
        "Hydrolysis": {
            "Data": 100,         // always 0-100
            "Unit": "%",         // always "%"
            "Setpoint": 50,      // always in %
            "Max": 100,          // always 100
            "System": {          // Only avail. on g/h systems
                "Data": 18, 
                "Unit": "%",
                "Setpoint": 18,
                "Max":33
            },
```

**Comment by alexdelprete** [comment]
> BTW: I have updated the [NeoPool documentation](https://tasmota.github.io/docs/NeoPool/) yesterday (added the [JSON key desc](https://tasmota.github.io/docs/NeoPool/#sensor-data-description) and [M5Stack usage](https://tasmota.github.io/docs/NeoPool/#using-m5stack-atom-lite-with-tail485-addon)).

Good job on the Tail485 doc. And you're talking with the FIRST one (self-pat on the back) who came up with that solution for SV/Hayward systems, simply because all that TTL conversion crap and all those components was too many things that could break. So I looked for a cleaner solution, and since I used Atoms for other projects, I checked M5Stack first, and it was incredible finding the Tail485. It worked immediately.

For the cable, I used the cable of a broken Hayward display because I didn't have the JST connector. :)

![image](https://github.com/alexdelprete/HA-NeoPool-MQTT/assets/7027842/f6289e29-bd26-43a1-ab3f-310b8537fd10)
![image](https://github.com/alexdelprete/HA-NeoPool-MQTT/assets/7027842/906b0409-3ef6-45eb-9e90-3938820f4c0a)

**Comment by curzon01** [comment]
> all that TTL conversion crap and all those components was too many things that could break. 

the biggest problem so far are always the converters, I've helped dozens of people having connection issues and it was always the same cause - some cheap 1-4 € RS485 converter from aliexpress etc..

> I used the cable of a broken Hayward display because I didn't have the JST connector.. :)

but isn't that a bit expensive if people have to buy a display and then break it? :grin:

**Comment by alexdelprete** [comment]
> No, this is a breaking change, suggestion:
> 
> ```json
>         "Hydrolysis": {
>             "Data": 100,         // always 0-100
>             "Unit": "%",         // always "%"
>             "Setpoint": 50,      // always in %
>             "Max": 100,          // always 100
>             "System": {          // Only avail. on g/h systems
>                 "Data": 18, 
>                 "Unit": "g/h",
>                 "Setpoint": 18,
>                 "Max":33
>             },
> ```

I understand, backwards compatibility issue.

Proposal: rename `System` to `GramsHourSystem` or just `GramsHour`.

Question: are you sure that Max value of the cell is needed only on g/h system? The SV electrolysis model has a cell that produces X g/h of chlorine, maybe a user wants to see it even if SV is set on % mode. Or am I missing something?

**Comment by alexdelprete** [comment]
> the biggest problem so far are always the converters, I've helped dozens of people having connection issues and it was always the same cause - some cheap 1-4 € RS485 converter from aliexpress etc..

Obviously, I noticed that crap even on amazon, not only on aliexpress, it's all made with low quality components. M5Stack makes great stuff and prices are great.

BTW: there's also this one, in an industrial packaging: [ATOM RS-485 Kit](https://shop.m5stack.com/products/atom-rs485-kit)
It contains the Atom Lite and instead of a tail device you connect the Atom on top via pin headers. It's more "bulky" but seems more robust, and you don't need to order two items, the kit contains both the Atom and the RS-485 device. HW components are the same.

**Comment by alexdelprete** [comment]
> but isn't that a bit expensive if people have to buy a display and then break it? 😁

You won't believe it, but they sent me a brand new Aquarite+ with the brand new touch display, and after 5s the display would turn off. :)

They wanted me to ship the Aquarite+ back, but I said: it's the display. And the seller said: How do you know? I said: I connected my Atom+Tail485 to the port of the display, and it works perfectly, power and communication are perfect. It's a defective display or cable.

So they sent me a replacement display and this one worked fine. I sent back the defective display, with a missing cable. 😁

**Comment by curzon01** [comment]
> Proposal: rename System to GramsHourSystem or just GramsHour.

I wanted a unitless key; also be aware of the key len (should be as short as possible due to ESP82xx limits)

> Question: are you sure that Max value of the cell is needed only on g/h system? The SV electrolysis model has a cell that produces X g/h of chlorine, maybe a user wants to see it even if SV is set on % mode. Or am I missing something?

Yes, on a % system you cannot evaluate the max g/h value, it is all in %. On my system (Oxilife 1) there are no g/h values of cell power in the official documents at all.

**Comment by curzon01** [comment]
> industrial packaging: [ATOM RS-485 Kit](https://shop.m5stack.com/products/atom-rs485-kit)

unfortunately EOL

**Comment by alexdelprete** [comment]
> I wanted a unitless key; also be aware of the key len (should be as short as possible due to ESP82xx limits)

`SysGH` - after all, there are only two possibilities, % or g/h.

> On my system (Oxilife 1) there are no g/h values of cell power in the official documents at all.

Because it's Hydrolysis + Electolysis. But for example, the Hidrolife has g/h on the display, it's just Electrolysis:

![image](https://github.com/alexdelprete/HA-NeoPool-MQTT/assets/7027842/996d631f-b1f6-4c8b-8b4a-6e96e516eca6)

> unfortunately EOL

Damn it, didn't know that. I wonder why...maybe the Tail485 replaced it?

**Comment by alexdelprete** [comment]
> I wanted a unitless key; also be aware of the key len (should be as short as possible due to ESP82xx limits)

other proposal: since it's a value specific for electrolysis cell, how about `Elysis` or `Cell`?

**Comment by alexdelprete** [comment]
> unfortunately EOL

The kit is EOL, the Base is not: https://shop.m5stack.com/products/atomic-rs485-base

**Comment by alexdelprete** [comment]
> I wanted a unitless key; also be aware of the key len (should be as short as possible due to ESP82xx limits)

Ok, I thought about it: I never understood why those g/h values were under the `Hydrolysis` json topic, since the chlorine production (measured in g/h) is based on Electrolysis, why don't we put those values under a new Electrolysis topic, at the same level of Hydrolysis?

Proposal:

```json
  "Hydrolysis": {
      "Data": 55,        // always 0-100
      "Unit": "%",       // always "%"
      "Setpoint": 55,    // always in %
      "Max": 100,        // always 100
      },
  "Electrolysis": {
      "Data": 18,         // always 0-100
      "Unit": "g/h",      // always g/h
      "Setpoint": 18,     // always 0-100
      "Max": 33,          // always 0-100
  }
```,

**Comment by curzon01** [comment]
>>On my system (Oxilife 1) there are no g/h values of cell power in the official documents at all.

> Because it's Hydrolysis + Electrolysis. But for example, the Hidrolife has g/h on the display, it's just Electrolysis

This is a common misconception.

Hydriolysis and electrolysis is the same thing. In all these systems, several titanium-coated plates operated with DC current are circulated with water containing sodium. The result is chlorine as well as hydrogen and two or three other chemical compounds present in smaller quantities.

Electrolysis is merely the chemical process leading to hydrolysis (the splitting of chemical compounds by reaction with water) - both terms are, in layman's terms, synonymous and, with respect to salt water electrolysis equipment, not different things.

The manufacturers like to promote with both terms in almost epic scale - it's all just marketing gossip

> I never understood why those g/h values were under the Hydrolysis json topic, since the chlorine production (measured in g/h) is based on the electrolysis process

The chlorine production with the "Chlorine" module has no chlore production measurement in g/h, it has only a currentl measured chlor value (Data) and an setpoint. And there is no extra electrolysis module either, internally it is all called hydrolysis, because hydrolysis and electrolysis is the same.

- The "Chlorine" module (the set is very expensive) contains besides a chlorine probe (that is the expensive one) also a chlorine pump.
The SV systems with "Chlorine" module do not produce chlorine but get liquid chlorine from a tank. It has only current value ("Data") and a setpoint, not production level.

- The SV systems with "Hydrolysis"/electrolysis module produce chlorine from the salt water.

Both modules ("Hydrolysis" and "Chlorine") can also be operated together, you can parameterize how which module should regulate.

The only thing that the "!Chlorine" module and the "Hydrolysis"/electrolysis module (parameterized with g/h instead of %) have in common is the unit. In one system, chlorine is produced (adjustable in % or g/h) and ORP is measured, in the other system chlorine is dosed and measured using a chlore probe.

Conclusion:

- There is no difference between hydrolysis and electrolysis in such devices, they are synonymous terms. Therefore, SV does not have an "electrolysis" module in the technical internal terms.
- The production level, set point and maximum value for a hydrolysis/electrolysis module (json key "Hydrolysis") are specified in g/h or % (depending on how the supplier has parameterized the device).
-  The data and setpoint for a chlorine controlled device (json key "Chlorine") does not have a production level, the level is based on pump performance alone and that is virtually unlimited.

Hope now it's more clear

**Comment by alexdelprete** [comment]
Thanks, I read about this process: actually they really distinguish among the two processes, at chemical level, at least it seems so. :)

![image](https://github.com/alexdelprete/HA-NeoPool-MQTT/assets/7027842/734605f2-ec83-45fa-935d-4de13fa527af)

**Comment by curzon01** [comment]
The difference is therefore only chemical. Both systems have the same titanium-coated cells.

If I have 3 g/l of salt in my water for my Oxilife, I have a hydrolysis system, if I have 1.5 g/l in the water, an electrolysis system - without having to buy a new device - magic

**Comment by curzon01** [comment]
Back to topic: 

The SENSOR topic for "Hydrolysis" will be kept unchanged but gets an additional key "Precent" which contains "Data" and "Setpoint" in percent regardless if it is a g/h or % system. On a g/h the values for the subkey "Precent" are converted accordingly.

```json
      // % system
      "Hydrolysis": {
          "Data": 40,
          "Unit": "%",
          "Setpoint": 75,
          "Max": 100,
          "Percent": {
            "Data": 40,
            "Setpoint": 75,
          },
          "Runtime": {
            "Total": "76T05:10:12",
            "Part": "76T05:09:55",
            "Pol1": "38T01:02:42",
            "Pol2": "38T04:07:30",
            "Changes": 562
          },
          "State": "Flow",
          "Cover": 0,
          "Boost": 0,
          "Low": 0
        },
```

```json
      // g/h system
        "Hydrolysis": {
          "Data": 13.2,
          "Unit": "g/h",
          "Setpoint": 24.7,
          "Max": 33,
          "Percent": {
            "Data": 40,
            "Setpoint": 75,
          },
          "Runtime": {
            "Total": "76T05:10:12",
            "Part": "76T05:09:55",
            "Pol1": "38T01:02:42",
            "Pol2": "38T04:07:30",
            "Changes": 562
          },
          "State": "Flow",
          "Cover": 0,
          "Boost": 0,
          "Low": 0
        },
```

I plan to append an optional % to the parameter for NPHydrolysis (`NPHydrolysis xxx %`), which will always set the value in %

This solution is backward compatible and the most flexible

**Comment by curzon01** [comment]
No add gh parameter. The old document was wrong but legacy user uses it in right way (as you, users with g/h machine simply adopt it's in g/h and users with % machine adopt its' in %, this will be kept as is and document will be update)

**Comment by alexdelprete** [comment]
> The old document was wrong but legacy user uses it in right way (as you, users with g/h machine simply adopt it's in g/h and users with % machine adopt its' in %, this will be kept as is and document will be update)

Now it makes sense. So this means that without specifying % as second parameter, it will work in legacy mode, interpreting the value based on the system, correct?

If I want an integration to always work in % mode, I am forced to use the second parameter.

**Comment by alexdelprete** [comment]
@curzon01 let me know when you implemented the changes in tasmota so I will recompile and make the needed changes to the integration.

**Comment by curzon01** [comment]
> So users could work in both modes if they want.

that's not possible on a % machine because you can't set a % machine in g/h. remember I said, the max g/h on a manufacturer delivered % machine is unknown

**Comment by alexdelprete** [comment]
> that's not possible on a % machine because you can't set a % machine in g/h. remember I said, the max g/h on a manufacturer delivered % machine is unknown

I wrote: "on g/h systems I could..." :)

**Comment by curzon01** [comment]
Alessandro, pls can you try Tasmota version having NPHydrlysis with precent value using your g/h system from my repo https://github.com/curzon01/Tasmota/tree/neopool ? 

It always returns value and unit:
```
NPHydrolysis 75
NPHYDROLYSIS = {"NPHydrolysis":75,"Unit":"%"}

NPHydrolysis 60 %
NPHYDROLYSIS = {"NPHydrolysis":60,"Unit":"%"}

NPHydrolysis 55%
NPHYDROLYSIS = {"NPHydrolysis":55,"Unit":"%"}
```

if it's fit I will make a Tasmota PR and add it to docs

desc from src
```
 * NPHydrolysis {<level> {%}}
 *            (only available if hydrolysis/electrolysis control is present)
 *            get/set hydrolysis/electrolysis level
 *            get current level if <level> is omitted, otherwise set:
 *            0..100 in % for systems configured to %
 *            0..<max> in g/h for systems configured for g/h (<max> depends by MBF_PAR_HIDRO_NOM register value)
 *            <level> can specified in % on all systems by appending the % sign to the value
```

**Comment by curzon01** [comment]
v13.2.0.1 Add NeoPool NPHydrolysis percent and unit 

[tasmota32-neopool.zip](https://github.com/alexdelprete/HA-NeoPool-MQTT/files/13251664/tasmota32-neopool.zip)

I have built in a small emulation for g/h, but fw should be tested on a real system

thx in advance for testing

**Comment by alexdelprete** [comment]
> thx in advance for testing

Seems to work fine. I tested with two different setpoints, one in % the other in g/h, and whichever you use, the other changes dynamically. But I will probably just leave the % mode.

![image](https://github.com/alexdelprete/HA-NeoPool-MQTT/assets/7027842/463b8155-0962-4fce-8915-a8e6d443677e)

**Comment by alexdelprete** [comment]
@curzon01 Norbert, for the sensor that displays the production (`Hydrolysis.Data`), since it could be a `g/h` system or a `percent` system, I had to prepare two sensors, since `unit_of_measurement` can't be "templateble" in HA. Since I didn't like that they were both active, I used the availability topic to make one of the two unavailable.

The only problem is that the lovelace file will have to be modified by the user based on the system.

It would've been nice to have one sensor dynamically configured based on the json, but it's not possible. This looks like a good compromise, let me know what you think.


```yaml
    - unique_id: "neopool_mqtt_hydrolysis_data_gh"
      name: "NeoPool MQTT Hydrolysis Data (g/h)"
      unit_of_measurement: "g/h"
      icon: mdi:water-opacity
      state_class: measurement
      state_topic: "tele/SmartPool/SENSOR"
      value_template: >-
        {{ value_json.NeoPool.Hydrolysis.Data | round(0) }}
      availability_topic: "tele/SmartPool/SENSOR"
      availability_template: >-
        {{ iif(value_json.NeoPool.Hydrolysis.Unit == "g/h", "Online", "Offline") }}
      payload_available: "Online"
      payload_not_available: "Offline"

    - unique_id: "neopool_mqtt_hydrolysis_data"
      name: "NeoPool MQTT Hydrolysis Data"
      unit_of_measurement: "%"
      icon: mdi:water-opacity
      state_class: measurement
      state_topic: "tele/SmartPool/SENSOR"
      value_template: >-
        {{ value_json.NeoPool.Hydrolysis.Data | round(0) }}
      availability_topic: "tele/SmartPool/SENSOR"
      availability_template: >-
        {{ iif(value_json.NeoPool.Hydrolysis.Unit == "%", "Online", "Offline") }}
      payload_available: "Online"
      payload_not_available: "Offline"
```

**Comment by curzon01** [comment]
Since a % system unfortunately cannot be converted into a g/h system, you always have either a valid g/h or None with an HA g/h sensor but % sensor is available on both systems. Suggestion:

If you want to keep both sensors use Hydrolysis.Percent.Data for % sensor (availability could be from same source as for all other standard sensors) and use Hydrolysis.Data as above for your g/h sensor.

**Comment by alexdelprete** [comment]
> If you want to keep both sensors use Hydrolysis.Percent.Data for % sensor (availability could be from same source as for all other standard sensors) and use Hydrolysis.Data as above for your g/h sensor.

Agreed. Will make the change. Thanks.

Did you submit your code to tasmota dev branch?

**Comment by curzon01** [comment]
> Did you submit your code to tasmota dev branch?

yes, it's merged

currently working on sending telemetry automatically when data changes in the system. This will solve the problem with the backlog construct.

**Comment by alexdelprete** [comment]
> currently working on sending telemetry automatically when data changes in the system. This will solve the problem with the backlog construct.

So my integration won't be needed anymore? Native integration with proper entity types? Great news, can't wait. :)

**Comment by curzon01** [comment]
> So my integration won't be needed anymore? Native integration with proper entity types? Great news, can't wait to test it. :)

No, unfortunately you misunderstood:

I meant that you don't need the construct "Backlog <cmnd>;Delay xx;Teleperiod" any more than you need an unnecessarily short teleperiod time. The TelePeriod SENSOR is then sent both teleperiod time and on immediate change of related values in the system.

**Comment by alexdelprete** [comment]
So SENSOR topic will be sent as scheduled by TelePeriod but also unsolicitedly, on every CHANGE of one of the values? That should work fine.

The only thing that worries me is why we needed that Delay before calling Teleperiod. I hope that with this improvement HA reacts immediately as we expect.

**Comment by curzon01** [comment]
BTW: The Tasmota HA integration is defined centrally in Tasmota ([Supported Entities](https://tasmota.github.io/docs/Home-Assistant/#supported-entities)). This cannot be extended driver-specifically. HA sensors for NeoPool are already implemented and working, they are also displayed for me:


![image](https://github.com/alexdelprete/HA-NeoPool-MQTT/assets/6636844/68338ab5-f39d-4eb1-85af-44c2c4a3ef55)

I think you mean a complete NeoPool-specific integration. That would be so much overkill that nobody would accept it, think of the ESP82xx systems.

What I am thinking about in the mid term is a HACS extension, simply click and install.

I have something like this for my heat pump, the MQTT source is a program on an ESP ([HeishaMon](https://github.com/Egyras/HeishaMon)) and the HACS extension ([heishamon-homeassistant](https://github.com/kamaradclimber/heishamon-homeassistant#heishamon-homeassistant)) brings this to HA.

You've been on HA for a while, take a look and see if you could adopt something similar.

**Comment by alexdelprete** [comment]
> This cannot be extended driver-specifically.

That's the problem. The default integration (Sensor Types) is not complete, and can't be extended/customized in the driver. But that's Tasmota, we can't do much about it, apart asking for a tighter integration or a more flexible model.

That's why [haco](https://github.com/fmtr/haco) was created I guess.



> I have something like this for my heat pump, the MQTT source is a program on an ESP ([HeishaMon](https://github.com/Egyras/HeishaMon)) sending [Topics](https://github.com/Egyras/HeishaMon/blob/master/MQTT-Topics.md) and the HACS Integration ([heishamon-homeassistant](https://github.com/kamaradclimber/heishamon-homeassistant#heishamon-homeassistant)) brings this to HA.
> 
> You've been on HA for a while, take a look and see if you could adopt something similar.

That would be the perfect solution: instead of using haco, a custom component for HA that reads MQTT and creates/updates all necessary sensors would be perfect.

I've [created](https://github.com/alexdelprete/ha-abb-powerone-pvi-sunspec) and I'm maintaining a custom component for my solar inverter, but I'm no real developer, I started from a component for another inverter and adapted and extended it.

I restarted programming after 20y only because of this component. :)

I could fork `heishamon-homeassistant` and use it as the base for a NeoPool component, hoping I'm up to the task. But I have to know you'll be there to support if/when needed, you're the real developer here. I'm just an old IT dinosaur that likes to tinker on many things.

Let me know...meanwhile I'll start forking and checking that component...
